test/distributions/sparseBlock/targLocsLocBlocks.chpl

```diff
@@ -0,0 +1,81 @@
+use BlockDist, LayoutCS;
+
+config const n = 10,
+             useTupleIndexing = true;
+
+const D = {0..<n, 0..<n} dmapped new blockDist({0..<n, 0..<n},
+                                               sparseLayoutType=CS(compressRows=true));
+const DS: sparse subdomain(D);
+
+// check that targetLocales on domains is working as expected
+writeln("DS target locales:");
+writeln(DS.targetLocales());
+writeln();
+
+var A: [DS] real;
+
+// check that targetLocales on arrays is working as expected
+writeln("A target locales:");
+writeln(A.targetLocales());
+writeln();
+
+// build up the global sparse domain and array by having each locale set
+// its own local blocks
+coforall loc in DS.targetLocales() do on loc {
+  // compute this locale's contribution to the index set, creating a
+  // diagonal pattern per locale
+  const myInds = DS.parentDom.localSubdomain();
+  var locSpsInds: sparse subdomain(myInds) dmapped new dmap(new CS(compressRows=true));
+
+  for i in 0..<min(myInds.dim(0).size, myInds.dim(1).size) {
+    locSpsInds += myInds.low + (i,i);
+  }
+  //  writeln(here.id, ": ", locSpsInds);
+
+  // now do the same for its local non-zeroes
+  var locVals: [locSpsInds] real;
+  for i in 0..<min(myInds.dim(0).size, myInds.dim(1).size) {
+    const low = myInds.low;
+    locVals[low+(i,i)] = low(0)+i + (low(1)+i)/10.0;
+  }
+  //  writeln(here.id, ": ", locVals);
+
+  // Take those local portions and plug them into the global domain/array
+  DS.setLocalSubdomain(locSpsInds);
```

Comment:
Will the array / domain relationship be inconsistent between these two calls? I think that this `setLocalSubdomain` will cause the local `myElems` array to be resized (right?) so it will not be inconsistent between these calls. However, it's doing more work than necessary, because instead of preserving the old data and resizing it, it could replace the domain and array at the same time. Might be worth a comment / Future Work item somewhere.

Reply:
Great point.  I believe you're right: that it won't be incorrect / inconsistent, but that it does involve wasted work for exactly the reason you point out.

We could potentially eliminate a good chunk of that work by setting the local subdomain to the empty domain prior to assigning the new domain…  That would eliminate the preservation of array elements, though it would still zero-iniitalize the arrays before re-assigning them with the new values.

So some sort of version that took the domain and array arguments simultaneously (or the domain and a list of array arguments?)  or used some sort of unsafe block/context manager [shudder] could lead to additional improvements.  

I may look into the "resize to zero" approach to see whether that is easy and lower overheads, but like the idea of adding a comment for the more impactful approaches for now.  I also suspect I'll bump back into this when I start doing more performance tuning of this code.

Tagging @e-kayrakli on this since it relates a bit to our discussion of higher-level interfaces for doing these sorts of assignments.

Reply:
> So some sort of version that took the domain and array arguments simultaneously (or the domain and a list of array arguments?)

I like this approach, FWIW

> like the idea of adding a comment for the more impactful approaches for now.

sounds good, it's all I was asking for here. Perhaps better in a Future Work section in the PR message vs a comment in this test.

Reply:
I was curious whether I could see a speedup if I used a `.clear();` call on the domain before re-assigning it, but couldn't really see any difference—though the execution times are very variable and very quick.  Digging into the code a bit more, I'm not convinced that `.clear()` is doing anything particularly reasonable for sparse, nor that domain-domain assignment is optimized for the case that the domains are the same type to just grab the vectors rather than adding indices piecemeal (though I'm not confident either way).  I was thinking I'd just add the `.clear()` in the "what harm could it do?" sense, but now that I've lost confidence that it's doing something reasonable, I think I'll take your future work approach idea instead.